### ISSUE TITLE
cargo fmt failing in Windows with MAX_PATH problem. can you add manifest with longPathAware?

### ISSUE THREAD

**Issue by kazuhiko-kikuchi** [rust-lang/rustfmt#5975]
longPathAware manifest item is documented here.

https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#longPathAware

and embedding manifest resource in build, can be with embed-resource crate. https://crates.io/crates/embed-resource

**Comment by ytmimi** [comment]
@kazuhiko-kikuchi thanks for reaching out. Any chance you can provide us with some more detail. What version of `cargo fmt` are you using? You can run `cargo fmt --version` to figure that out. Additionally, can you put together a minimal reproducible test case, though if this is an windows OS issue then I don't know if there's much we can do on the rustfmt side.

**Comment by kazuhiko-kikuchi** [comment]
```
$ cargo fmt --version
rustfmt 1.6.0-stable (79e9716c 2023-11-13)
```

We use deep directory structure. `cargo fmt` succeeded in Linux, but Windows in fails. 

```
$ cargo fmt
ファイル名または拡張子が長すぎます。 (os error 206)
This utility formats all bin and lib files of the current crate using rustfmt.

Usage: cargo fmt [OPTIONS] [-- <rustfmt_options>...]

Arguments:
<snip>
```

`ファイル名または拡張子が長すぎます。 (os error 206)` is os error message for MAX_PATH problem.

**Comment by PatchMixolydic** [comment]
> ファイル名または拡張子が長すぎます。 (os error 206)

To aid investigation, the English version of this error message [seems to be](https://answers.microsoft.com/en-us/windows/forum/all/error-206-while-trying-to-start-windows-search/ae93193b-95f1-4d9c-8936-a7fb4c060e32) "The file name or extension is too long."

**Comment by ytmimi** [comment]
@kazuhiko-kikuchi are you able to link to the project where you're trying to run `cargo fmt`? If not, are you able to put together a minimal test case that we can use to reproduce the issue that you're running into.

**Comment by RossSmyth** [comment]
@ytmimi This was actually found with rust-lang/miri#3317 recently as well. It is most likely unrelated to MAX_PATH. Though it possible for that to be an issue.

The issue is the following:

1. The miri repo runs rustfmt on every `.rs` file in the repo. This includes tests and every crate.
2. [Windows has a hard limit on the length of CLI arguments + environment variables of 32k](https://devblogs.microsoft.com/oldnewthing/20031210-00/?p=41553)
3. [The way miri's formatting tool currently works is collecting every file in the repo and throwing it at rustfmt](https://github.com/rust-lang/miri/blob/40bb3c110b2ceaf9f8b134c99aa05282d81b77f3/miri-script/src/commands.rs#L533-L559)
4. This leads to the formatting tool not working on Windows, as the total CLI argument length is around 65k

It looks like cargo-fmt is also susceptible to this issue as well.
https://github.com/rust-lang/rustfmt/blob/21f353a6373126035f9063780bdccb9dd191a2bb/src/cargo-fmt/main.rs#L507-L519

I have a fix up for the miri tool, which batches rustfmt calls by argument length.
https://github.com/rust-lang/miri/blob/c72e73f4deab94cb433e64b4c863124eb893ef6c/miri-script/src/util.rs#L153-L229

Basically the "minimal" test case is:
1. Have a Windows machine
2. Generate a whole bunch of files to throw at rustfmt so that the arguments + env var length goes above (2^15 - 1)
3. Run cargo-fmt

Other OS's have limits as well, but they are much higher.

**Comment by kazuhiko-kikuchi** [comment]
I know cargo have longPathAware manifest now.

https://github.com/rust-lang/cargo/issues/7986

**Comment by RossSmyth** [comment]
Yeah, though cargo-fmt is a seperate binary. Since you have not provided a minimal reproducer there are potentially two different issues here:

1. It could be longPathAware
2. It could be passing too many characters via CLI

They have the same error from Windows it is impossible to disambiguate them from the information provided. But both should be mitigated as they are both issues.

**Comment by kazuhiko-kikuchi** [comment]
> Since you have not provided a minimal reproducer

sorry, I work on closed source repo. Can not to provide it.

**Comment by LikeLakers2** [comment]
I'm experiencing this issue with the https://github.com/bevyengine/bevy repository. Running `cargo fmt` returns the following error:

> The filename or extension is too long. (os error 206)

This only occurs if the repository is in a directory whose path (including the trailing `\`) is 46 characters long. For example, `C:\Users\admin\Documents\Repos\Forks\bevy\` is 42 characters long, which is short enough - but `C:\Users\admin\Documents\Repos\Forks\bevyengine.bevy\` is 53 characters long, making it too long.

**Comment by RichardLexJiang** [comment]
I'm experiencing same issue , mark .

**Comment by IceSentry** [comment]
I just had the same issue with bevy like @LikeLakers2 explained. I had to move my repo at a location that is shorter and now it works.

**Comment by RossSmyth** [comment]
Once again this is an issue with CLI arg length.
